### ISSUE TITLE
Mac: App: open-file event isn't fired

### ISSUE THREAD

**Issue by adam-lynch** [atom/electron#4403]
I've the following at the top of my main JS file;

```javascript
'use strict';
const electron = require('electron');
const app = electron.app;
app.on('open-file', function(event, filePath){
  event.preventDefault();
  console.log(filePath);
});
```

It isn't fired when I open the app by dragging a file onto the icon and nothing happens when I drag & drop a file on the dock icon.

**Comment by adam-lynch** [comment]
Ignoring my stupidity (trying to use `console.log` in the main process)... I think this is still valid.

It's pretty hard to debug this but I've managed to prove that `open-file` is fired if the app is opened with a file. So great. That leaves dropping a file on the dock icon after the app is already running... nothing is happening. In a packaged or unpackaged (dev) app.

**Comment by kevinsawicki** [comment]
@adam-lynch does the icon in the dock change color when you drag the file over it?

**Comment by adam-lynch** [comment]
@kevinsawicki I'm not at that machine right now but I think it didn't.

**Comment by erkyrath** [comment]
I got file-open working on MacOS.

Two things. First, I set the 'open-file' event handler inside my 'will-finish-launching' handler. That is, I have code like this:

	app.on('will-finish-launching', function() {
	    app.on('open-file', function(ev, path) {
	        //...
	    });
	});

It's probably okay to set it at the top level, but I haven't tested it that way.

Second, I added a CFBundleDocumentTypes section to Mac app's Info.plist and listed the acceptable file types. (See https://developer.apple.com/library/ios/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html )

**Comment by kevinsawicki** [comment]
> Second, I added a CFBundleDocumentTypes section to Mac app's Info.plist and listed the acceptable file types.

@adam-lynch does your app have this :point_up: ? I think it might be required.

**Comment by adam-lynch** [comment]
@kevinsawicki no I don't. Looking into it now. 

@erkyrath do you use electron-packager? How are you adding the info.plist?

I see what I've to put in the info.plist but not sure where to put the info.plist (I think in Resources?) or what exactly I can get away with leaving out from this example: https://github.com/atom-archive/electron-starter/blob/master/resources/mac/app-Info.plistt.

**Comment by erkyrath** [comment]
Start with the Electron.app/Contents/Info.plist in electron-prebuilt and add to it.

I use electron-packager. To add to the Info.plist I had to write some extra code -- see PR https://github.com/maxogden/electron-packager/pull/253.

My project is at https://github.com/erkyrath/lectrote , you can see how I did it. Look at my package.json.

**Comment by zcbenz** [comment]
To make `open-file` event work you have to make a packaged app, @erkyrath has covered all other things necessary.

**Comment by adam-lynch** [comment]
Thanks @erkyrath. 

---

Ok, there's definitely something wrong here. First, let's look at what the docs say about the `open-file` event:

> Emitted when the user wants to open a file with the application. The open-file event is usually emitted when the application is already open and the OS wants to reuse the application to open the file. open-file is also emitted when a file is dropped onto the dock and the application is not yet running. Make sure to listen for the open-file event very early in your application startup to handle this case (even before the ready event is emitted).

Here are my findings (using `will-finish-launching` by the way): 

# No plist changes

## Finder

#### Not running 

I can drop any type of file on the app in finder: the app icon will show a gray background behind it, it will open the app, and emit the `open-file` event. 

#### Already running

Dropping the file on the app in finder when the app is already running results in the app coming to the front but the `open-file` event is not emitted.

## Dock

#### Not running

Nothing happens.

#### Already running

Nothing happens. No visual change to the icon. The app is not brought to the front.

## Open with...

I.e. right clicking a file, clicking Open With, then Other, navigating to the directory where the app is.

#### Not running

The app is shown as a "Recommended Application" (not gray). Selecting the app will open it and then emit `open-file`.

#### Already running

The app is shown as a "Recommended Application" (not gray). Selecting the app will bring the appto the front but `open-file` won't be emitted.

# With plist changes

I've used @erkyrath's electron-packager fork to achieve this. The plist in the app is:

```plist
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>BuildMachineOSBuild</key>
    <string>14B25</string>
    <key>CFBundleDisplayName</key>
    <string>electron-quick-start</string>
    <key>CFBundleExecutable</key>
    <string>Electron</string>
    <key>CFBundleIconFile</key>
    <string>atom.icns</string>
    <key>CFBundleIdentifier</key>
    <string>com.electron.electron-quick-start</string>
    <key>CFBundleInfoDictionaryVersion</key>
    <string>6.0</string>
    <key>CFBundleName</key>
    <string>electron-quick-start</string>
    <key>CFBundlePackageType</key>
    <string>APPL</string>
    <key>CFBundleShortVersionString</key>
    <string>0.36.7</string>
    <key>CFBundleVersion</key>
    <string>0.36.7</string>
    <key>DTSDKBuild</key>
    <string>14D125</string>
    <key>DTSDKName</key>
    <string>macosx10.10</string>
    <key>DTXcode</key>
    <string>0640</string>
    <key>DTXcodeBuild</key>
    <string>6E35b</string>
    <key>LSApplicationCategoryType</key>
    <string>public.app-category.developer-tools</string>
    <key>LSMinimumSystemVersion</key>
    <string>10.9.0</string>
    <key>NSHighResolutionCapable</key>
    <true/>
    <key>NSMainNibFile</key>
    <string>MainMenu</string>
    <key>NSPrincipalClass</key>
    <string>AtomApplication</string>
    <key>NSSupportsAutomaticGraphicsSwitching</key>
    <true/>
    <key>CFBundleTypeExtensions</key>
    <array>
      <string>png</string>
    </array>
    <key>LSItemContentTypes</key>
    <array>
      <string>public.png</string>
    </array>
  </dict>
</plist>
```

Note the last two keys; CFBundleTypeExtensions (deprecated as of Mac OS X 0.10.5) and LSItemContentTypes (which overrides CFBundleTypeExtensions if it exists, see https://developer.apple.com/library/ios/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html).

## Finder

#### Not running 

Same as without any plist changes.

#### Already running

Same as without any plist changes.

## Dock

#### Not running

Same as without any plist changes.

#### Already running

Same as without any plist changes.

## Open with...

#### Not running

Same as without any plist changes (with a PNG). What's noteworthy here is that a PDF also works even though I've only listed PNG as a supported file type in the Info.plist.

#### Already running

Same as without any plist changes (with any file type including PNGs). 

---

My setup:

Mac OS X: 10.11.1
Electron: 0.36.7

---

@kevinsawicki @zcbenz So I think maybe there are bugs. Or else the docs are wrong. Or if I'm wrong (please tell me what I'm doing wrong) and CFBundleTypeExtensions & LSItemContentTypes are indeed needed, then it should be documented.

**Comment by zcbenz** [comment]
Electron is just passing the information got from OS X, and the document just replicates what's on Apple's website: 
https://developer.apple.com/library/mac/documentation//Cocoa/Reference/NSApplicationDelegate_Protocol/index.html?ref=driverlayer.com/web#//apple_ref/occ/intfm/NSApplicationDelegate/application:openFile:

**Comment by adam-lynch** [comment]
Ok thanks @zcbenz. Maybe something should be in the docs though.

**Comment by johanpoirier** [comment]
@adam-lynch I still have the issue you described: 

> Open with...
>
>Already running
>
>Same as without any plist changes (with any file type including PNGs).

Did you manage to get it working?

**Comment by adam-lynch** [comment]
@johanpoirier No, I don't think so.

**Comment by johanpoirier** [comment]
So why is the issue closed @zcbenz ?

**Comment by zcbenz** [comment]
All other apps like Atom and VS Code are using this event fine, so I think it is more likely user's fault.

**Comment by johanpoirier** [comment]
OK but it's not working for me either.

I have an app with a single instance:
 - app is closed, I open a file with "Open with": the `open-file` event is fired
 - app is already opened and I try to open another file with "Open with": no event
 - app is already opened and I drop a file on the dock icon: the `open-file` event is fired

I tried to use the `argv` from the `app.makeSingleInstance` callback but the file path is not in the arguments.

So what am I doing wrong?

**Comment by johanpoirier** [comment]
Hello, any update on this?

**Comment by antonycourtney** [comment]
I am also encountering similar behavior to @adam-lynch on macOS Sierra, where the `open-file` event is simply not being delivered.  Can we please re-open?

I have an application that calls `app.makeSingleInstance` and registers an `open-file` event handler.  I'm using [electron-builder](https://github.com/electron-userland/electron-builder), and have a [file association](https://github.com/electron-userland/electron-builder) for CSV files that I've verified results in the appropriate entries in `Info.plist`.

I am primarily interested in being able to open CSV files with my via "Open With..." from the Context menu in the Finder. What I find happens is this:

- If my application is not running, it is started and the `open-file` event for the selected file is delivered to the application, as I would expect.  This, combined with visual inspection of the Info.plist file (and comparison to the corresponding entries in Info.plist for `.csv` file handling in Atom), leads me to believe that the file association information in `Info.plist` is correct.

- If there is already an instance of the application running, the callback for `makeSingleInstance` in the running app will be invoked, but **the `open-file` event is not delivered to the running instance of the application.**.  The `argv` passed to the `makeSingleInstance` callback invocation seems odd and not very useful; it is something like:
```
[ '/Volumes/Tad 0.6.6/Tad.app/Contents/MacOS/Tad',
  '-psn_0_1503599' ]
```
I have no idea what's going on with `argv[1]` (`-psn_0_...`) there.

The rationale @zcbenz gave for closing this was "works for Atom...must be the user's fault", which seemed a bit premature to me. 
Quick question:  Does Atom use `app.makeSingleInstance`?  From my (very cursory) reading of the Atom source it looks to me like maybe it has its own socket-based mechanism that it uses instead.  I think the issue here is the interaction of `open-file` with `makeSingleInstance`.

The really painful part about issues like this is that, as far as I know, one has to build and install a fully packaged app to actually test (which for me takes about 5 to 10 min), and it's much more difficult to see what's going on in a packaged app than an unpackaged dev one.  If there is interest and willingness to re-open I'm willing to try to put together a minimal reproducible example.

**Comment by mediaslav** [comment]
you can place symlink to "app" dir inside installed app, replacing app.asar, and save (re)building time

**Comment by antonycourtney** [comment]
@mediaslav Thanks for that tip!  Hadn't known that; will indeed be a huge time saver for debugging this sort of issue.

**Comment by antonycourtney** [comment]
@zcbenz @kevinsawicki Can we please re-open this issue, per my comment above?  I've just had to revert my app from calling `app.makeSingleInstance()` because I don't get open-file events if I do. I will be happy to help provide additional detail or work on a minimal repro if that would be helpful.

**Comment by kevinsawicki** [comment]
@antonycourtney can you open a new issue with reproduction steps for what you are seeing?

The original report here did not involve `app.makeSingleInstance` and was a packaging issue so it would be better to open a new issue that is more explicit to the issue you are seeing, thanks.

**Comment by antonycourtney** [comment]
@kevinsawicki Sure thing, will do. Thanks.

**Comment by HemalR** [comment]
@antonycourtney 

Was there a newer updated issue for this? Been banging my head on the wall for two days trying to get the open-file event working and all the examples I can find have just the code:

```
app.on('will-finish-launching', ()=>{
  app.on('open-file', (event, path)=>{
    event.preventDefault();
    fileToOpen = path;
  
    if (mainWindow) {
      mainWindow.send('open-file', path);
    }
  })
})
```

Which is simply not working 😕 

I'm testing packaged apps and the dock icon isn't changing colour when dragging a file on top of it either.

**Comment by jakesyl** [comment]
So, I guess this doesn't work?

**Comment by HemalR** [comment]
@jakesyl 

Specifically for Mac - I figured out the issue (my apologies for not updating this). I used electron-builder to tell the system that a certain type of file can be opened by the program:

In `package.json` under the `build` key:
```
"mac": {
			"category": "public.app-category.business",
			"type": "distribution",
			"fileAssociations": [
				{
					"ext": ["jpg", "jpeg", "png", "PNG", "tiff"],
					"description": "Image files",
					"role": "Viewer"
				},
				{
					"ext": ["pdf", "PDF"],
					"description": "PDF files",
					"role": "Viewer"
				}
			]
		},
```

That did the trick!

**Comment by ckerr** [comment]
@HemalR, thank you for the followup! Hopefully it will leave some breadcrumbs for future web searchers :+1:

**Comment by jakesyl** [comment]
Awesome @HemalR. Thanks for the follow up. Anyone struggling with this same issue, I suggest you checkout #9925!

**Comment by payitforwardnow** [comment]
I'm having this exact same problem. I added some comments to #9925 since both are closed. I've verified that the fileAccociations configuration is correct and checked the info.plist file that is created to verify. I've put the open-file event hook before AND after the app is ready. I've been able to determine that the open-file event fires as expected ONLY if the app is launched a specific way. See the comments in #9925 for anyone that is having the same problem. Thank you in advance for your time and consideration.
